### PR TITLE
[Fix #3838] Add RuboCop::Cop::Badge

### LINKED ISSUE #3838: CopStore assumes that all cop departments have one capital letter
I noticed this while working on #3887 and figured I would open an issue even though I'm working on a fix. The `CopStore` assumes that you can downcase all cop namespaces and then capitalize them to get the original namespace. See https://github.com/bbatsov/rubocop/blob/ffe9743b1d8c4b37f04041022ec6fb388c04e531/lib/rubocop/cop/cop.rb#L29.

This works for all of rubocop's department but not for rubocop-rspec:

| Original Department   | Downcased     | Capitalized   |
|:---------------------|:--------------|:--------------|
| `Bundler`            | `bundler`     | `Bundler`     |
| `Lint`               | `lint`        | `Lint`        |
| `Metrics`            | `metrics`     | `Metrics`     |
| `Performance`        | `performance` | `Performance` |
| `Style`              | `style`       | `Style`       |
| `Rails`              | `rails`       | `Rails`       |
| `Security`           | `security`    | `Security`    |
| `RSpec`              | `rspec`       | `Rspec`       |

As a result, writing `rubocop:disable IndentArray` works but `rubocop:disable AnyInstance` (rubocop-rspec cop) does not.

Expected behavior would be to properly inflect `:rspec` back to `:RSpec`.

### PR DESCRIPTION
**Pull request by backus**
As mentioned in #3838, all cops expose a `#department` as a downcased symbol even though these departments seem to always be inflected (using `#capitalize`) at later points to construct qualified cop names or other end user visible messages.

There are also many parts of the codebase that duplicate the behavior and rules for parsing and constructing cop names (`Department/CopName`):

This PR adds `RuboCop::Cop::Badge` which unifies this duplicated behavior. It also eliminates the unnecessary downcasing and later capitalizing of departments which fixes the issue mentioned in #3838. 

This was originally a part of #3837 but I've extracted part of the refactor to help reduce the size of the cop store PR.

-----------------

Duplication before and after

### Before

```sh
$ ag "split\('/"
lib/rubocop/config.rb
326:        next unless key?(cop_name) || key?(cop_name.split('/').last)

lib/rubocop/cop/lint/unneeded_disable.rb
162:            matching_range(comment.loc.expression, cop.split('/').last) ||

tasks/cops_documentation.rake
9:    cop_name.split('/').last.to_sym

tasks/new_cop.rake
15:  category, name = *cop_name.split('/', 2)

$ ag "join\('/"
lib/rubocop/cop/cop.rb
107:        @cop_name ||= name.split('::').last(2).join('/')

$ ag "}/\#\{"
lib/rubocop/cop/cop.rb
30:          cop_names.include?("#{ns}/#{basename}")
39:                   "#{found_ns.map { |ns| "#{ns}/#{basename}" }.join(' or ')}"
48:        "#{found_ns}/#{basename}"

$ ag '(department|type).+capitalize'
lib/rubocop/cli.rb
127:        puts "# Department '#{department.to_s.capitalize}' " \

lib/rubocop/config.rb
145:      department = cop.department.to_s.capitalize!

lib/rubocop/cop/cop.rb
29:        found_ns = departments.map(&:capitalize).select do |ns|
124:          given_names.include?(department.to_s.capitalize)

lib/rubocop/options.rb
182:      departments = Cop::Cop.all.departments.map { |t| t.to_s.capitalize }

spec/rubocop/cli/cli_options_spec.rb
596:          expect(stdout).to include(department.to_s.capitalize)
604:            department.to_s.capitalize

tasks/cops_documentation.rake
83:    content = "# #{type.capitalize}\n".dup
```

### After

```
$ ag "split\('/"
lib/rubocop/cop/badge.rb
24:        parts = identifier.split('/', 2)

$ ag "join\('/"
$ ag "}/\#\{"
lib/rubocop/cop/badge.rb
50:        qualified? ? "#{department}/#{cop_name}" : cop_name

lib/rubocop/cop/cop.rb
30:          cop_names.include?("#{ns}/#{basename}")
39:                   "#{found_ns.map { |ns| "#{ns}/#{basename}" }.join(' or ')}"
48:        "#{found_ns}/#{basename}"

spec/rubocop/ast_node_spec.rb
343:      let(:src) { "/#{body}/#{opts}" }

$ ag '(department|type).+capitalize'
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: lib/rubocop/cop/badge.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    # Identifier for all cops
```
Comment (bbatsov) [inline lib/rubocop/cop/badge.rb]: I'd extend this description a bit. Use sentences ending with `.`.

File: lib/rubocop/cop/badge.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    class Badge
+      # Error raised when a badge parse fails
+      class InvalidBadge < Error
+        MSG = 'Invalid badge %<badge>p. ' \
```
Comment (bbatsov) [inline lib/rubocop/cop/badge.rb]: Seems this `p` is here by accident.

**Comment by bbatsov** [comment]
Great refactoring overall! Just one question - will the changes affect passing a department name as a command-line option? E.g. to just run the cops from one department.

File: lib/rubocop/cop/badge.rb (hunk 2 above)
Comment (backus) [inline lib/rubocop/cop/badge.rb]: The `p` here means that I want `format` to call inspect. In other words `format("Value: %<name>s", name: 'John'.inspect)` is equivalent to `format("Value: %<name>p", name: 'John')`

File: lib/rubocop/cop/badge.rb (hunk 1 above)
Comment (backus) [inline lib/rubocop/cop/badge.rb]: Will do

**Comment by backus** [comment]
@bbatsov no it shouldn't. I believe those go through `cop_store.with_department(department_name)` which isn't touched here. I also double checked manually that I can do things like `bin/rubocop --only Bundler` and it works as expected.

**Comment by backus** [comment]
> Great refactoring overall!

Thanks :smile:

**Comment by backus** [comment]
@bbatsov updated. I didn't change the `%<badge>p` but I provided an explanation in the comments

File: lib/rubocop/cop/badge.rb (hunk 2 above)
Comment (bbatsov) [inline lib/rubocop/cop/badge.rb]: Ops, my bad. I'm not paying enough attention early in the morning. :-)

**Pull request merged**